compiler/fir/tree/src/org/jetbrains/kotlin/fir/types/FirTypeUtils.kt

```diff
@@ -165,12 +168,18 @@ val ConeKotlinType.canBeNull: Boolean
         return when (this) {
             is ConeFlexibleType -> upperBound.canBeNull
             is ConeDefinitelyNotNullType -> false
-            is ConeTypeParameterType -> this.lookupTag.typeParameterSymbol.fir.bounds.any { it.canBeNull }
-            is ConeIntersectionType -> intersectedTypes.any { it.canBeNull }
+            is ConeTypeParameterType -> this.lookupTag.typeParameterSymbol.allBoundsAreNullable()
+            is ConeIntersectionType -> intersectedTypes.all { it.canBeNull }
+            is ConeTypeVariableType -> lookupTag.originalTypeParameter?.safeAs<ConeTypeParameterLookupTag>()?.typeParameterSymbol?.allBoundsAreNullable()
```

Comment:
@demiurg906 I considered delegating [`ConeTypeContext#KotlinTypeMarker.isNullableType`](https://github.com/JetBrains/kotlin/blob/b59ac5d8f651ef2084705aa0f2cf4b63720f3cdc/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeContext.kt#L406-L425) to this function, however it was not yielding the same result for `ConeTypeVariableType`.

In fact I noticed that in that function, `lookupTag.toSymbol(session)` was always returning false. And also, if I comment out the entire `is ConeTypeVariableType` branch, the FIR tests all pass.

Any idea why this is so?

Reply:
Actually, nullability of type variable is more or less meaningless, because it's unknown until all constraints will be collected and variable will be fixed (but after that there won't be any type variable). So it's defined that all type variables are not null. Check FE 1.0 [counterpart](https://github.com/JetBrains/kotlin/blob/master/compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/inference/model/TypeVariable.kt#L74)

Reply:
Makes sense. However in the diagnostic, the expected type is still a type variable which is not fixed. Does what I added here in `FirTypeUtils.canBeNull` look reasonable and correct?

Reply:
The case which has a type variable for expected type in `InapplicableWrongReceiver` diagnostic is `useAsReceiver.kt` (excerpt:)

```kt
fun <T : CharSequence> T.bar3() {}
fun <T : String?> foo(x: T) {
    //...
    x<!UNSAFE_CALL!>.<!>bar3()
}
```

Reply:
Actually, better solution is to fix type variable to something more reasonable before reporting diagnostic (e.g. to corresponding type parameter type). After that changes in `canBeNull` for type variable types are redundant

Main rule about type variables is that variables should exist only during resolution/inference, and should not leak to other stages

Reply:
I agree it should be fixed before then, but I couldn't figure out where and how to do this correctly. Could you point me in the right direction, maybe in `checkApplicabilityForArgumentType()`?

Reply:
It's fine when `NewConstraintError` (or other resolution diagnostic) has type variable inside. And it should converted to proper type parameter type during convertion of `ResolutionDiagnostic` to `ConeDiagnostic`

Reply:
@demiurg906 - @tgeng has a separate PR and [one of its commits](https://github.com/JetBrains/kotlin/pull/4287/commits/ef87b0e42a70b5a4062e4e34c5b7c21aca45b93b) will render this PR obsolete by adding a `UnsafeCall` diagnostic separate from `InapplicableWrongReceiver`. If you think that's a reasonable approach then I'm happy to just drop this PR.

Reply:
Actually I would just drop the 1st commit. The 2nd commit (using intersection of upper bounds) is still valid.